File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/commands/RedeemGiftCard.java
package it.giannibombelli.wsc2026.giftcard.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;

public record RedeemGiftCard(GiftCardId aggregateId, Money amount) implements Command<GiftCardId> {
    public RedeemGiftCard {
        Require.requireArgument(aggregateId, "giftCardId");
        Require.requireArgument(amount, "amount");
    }
}

File: typescript/src/giftcard/application/commands/redeemGiftCard.ts
// Command per riscattare una gift card.

import type { Command } from '@/common/application/command.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type RedeemGiftCard = Command<'RedeemGiftCard', GiftCardId> & {
  readonly amount: Money;
};

export function redeemGiftCard(aggregateId: GiftCardId, amount: Money): RedeemGiftCard {
  requireArgument(aggregateId, 'giftCardId');
  requireArgument(amount, 'amount');
  return { kind: 'RedeemGiftCard', aggregateId, amount };
}
